src/library/scala/collection/mutable/ArrayBuilder.scala

```diff
@@ -31,16 +31,29 @@ sealed abstract class ArrayBuilder[T]
 
   override def knownSize: Int = size
 
+  private final val MaxArraySize:Int = 2147483639
+
   protected[this] final def ensureSize(size: Int): Unit = {
     if (capacity < size || capacity == 0) {
-      var newsize = if (capacity == 0) 16 else capacity * 2
-      while (newsize < size) newsize *= 2
-      resize(newsize)
+      if (size <= 1073741824) {
+        if (size < 0) throw new Exception(
+          s"ArrayBuilder encountered negative size parameter: $size, possibly as a result of overflow."
+        ) else {
+          var newSize = if (capacity == 0) 16 else capacity * 2
+          while (newSize < size) newSize *= 2
+          resize(newSize)
+        }
+      } else if (capacity < MaxArraySize && size <= MaxArraySize) {
+        resize(MaxArraySize)
+      } else throw new Exception(
+        s"ArrayBuilder cannot store more than $MaxArraySize elements and cannot ensure a size of $size."
+      )
```

Comment:
I would prefer if we could align the code here with https://github.com/scala/scala/blob/v2.13.11/src/library/scala/collection/mutable/ArrayBuffer.scala#L321-L325. It makes maintenance much easier in the long turn.

The implementation in `ArrayBuffer` doubles the current size, but if that's still too small it jumps to the exact requested size. Discussion about this is here: https://github.com/scala/bug/issues/12464#issuecomment-970841156.

Reply:
I can work on that.  You've linked to private methods; do you have a suggestion for where the shared logic should live?

Reply:
I don't have strong opinions, you could simply change `resizeUp` to be `private[mutable]` and leave it where it is in `object ArrayBuffer`.

Reply:
I did not intend to ignore this PR, I was waiting for dust to settle. I'll contribute eyeballs after I unbreak unused imports on my own PR.

Reply:
> I don't have strong opinions, you could simply change `resizeUp` to be `private[mutable]` and leave it where it is in `object ArrayBuffer`.

I'll try that.

Reply:
Are we sure that won't cause any binary incompatibility?

Reply:
Binary compat is about public API, what users might link, exclusive of inlining.

Reply:
@lrytz, I have concerns about the way `ArrayBuffer` handles resizing.

```scala
if (targetLen > Int.MaxValue) throw new Exception(s"Collections cannot have more than ${Int.MaxValue} elements")
IterableOnce.checkArraySizeWithinVMLimit(targetLen.toInt) // safe because `targetSize <= Int.MaxValue`
```
Why not just replace these two lines with the relevant internals of `checkArraySizeWithinVMLimit`:
```scala
if (targetLen > scala.runtime.PStatics.VM_MaxArraySize) throw new Exception(
  s"Size of array-backed collection exceeds VM array size limit of ${scala.runtime.PStatics.VM_MaxArraySize}"
)
```

The current way seems overly complicated and implies that the user can allocate an `Array`, `ArrayBuffer`, or `ArrayBuilder` of size `Int.MaxValue` when the true limit, `scala.runtime.PStatics.VM_MaxArraySize`, is smaller.

More importantly, anything involving `Long` will cause significant performance problems in Scala.js.  Since every single call to `addAll` or `addOne` will perform several operations on Long, this approach might be a mistake.  Maybe we ought to detect overflow without casting sizes to long values?

Reply:
In Scala.js, `Long`s are indeed expensive when they cross method boundaries. If you use `Long`s locally within a method they're OK.

Regardless, if you're using `Long`s only to test that the result of `a + b` does not overflow and you know that `a >= 0 && b >= 0`, the most efficient remains to test whether the result is `< 0`.

Reply:
@lrytz I'd like to follow sjrd on this and rewrite the long values out of the resizing logic.

Would you oppose?

Reply:
No, sounds good to me 👍

Reply:
When using doubling to come up with a new size, you can be careful with your operations to avoid overflow.  For instance, instead of testing `2*n >= maxSize` you can test `n >= maxSize/2`.  Or instead of `n + m >= maxSize` you can use `m >= maxSize - n` (where `n` is known to be between `0` and `maxSize`).

Once you do the check that it's in range, you still have to calculate the correct value (`2*n` or `n+m`), but integer math is very inexpensive, so that's no big deal.  For a core library routine, spending the effort to do it this way (correctly) is probably worth it.

(Another option is to do the `2*n` or `n + m` but then the test is `x >= maxSize || x < 0` to catch the overflow case, assuming that your input values are nonnegative.)